### PR TITLE
feat(upgrade): refactor upgrade controller and kubectl plugin

### PR DESCRIPTION
**Pull request by niladrih**
Upgrade operator changes:
- run controller asynchronously/respond to apply upgrade REST call
- add check to see if the io-engine pods all come to running state after upgrade
- add check to verify if all volumes are unpublished before restarting io-engine
- add loop to retry data-plane upgrade
- add check to verify if io-engine pods are restarted

Kubectl plugin changes:
- update upgrade operator deployment args flags
- add get, list, create, delete, patch permissions for poddisruptionbudget resource

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

### CODE REVIEW

File: operators/src/upgrade/common/constants.rs [hunk 1]
```diff
@@ -3,6 +3,7 @@ use std::{
     env,
     path::{Path, PathBuf},
 };
+use tracing::info;
```
Comment (tiagolobocastro) [inline operators/src/upgrade/common/constants.rs:6]: not used?

File: operators/src/upgrade/controller/reconciler.rs [hunk 2]
```diff
@@ -416,33 +451,37 @@ async fn reconcile(ua: Arc<UpgradeAction>, ctx: Arc<ControllerContext>) -> Resul
     }
 }
 
+pub(crate) enum Command {
```
Comment (tiagolobocastro) [inline operators/src/upgrade/controller/reconciler.rs:454]: missing doc comment

File: operators/src/upgrade/controller/reconciler.rs [hunk 3]
```diff
@@ -416,33 +451,37 @@ async fn reconcile(ua: Arc<UpgradeAction>, ctx: Arc<ControllerContext>) -> Resul
...
+            .run(reconcile, error_policy, Arc::new(context))
+            .for_each(|res| async move {
+                match res {
+                    Ok(o) => {
+                        trace!(?o);
+                    }
+                    Err(e) => {
+                        trace!(?e);
+                    }
+                }
```
Comment (tiagolobocastro) [inline operators/src/upgrade/controller/reconciler.rs:482]: trace!(?res) ?

File: operators/src/upgrade/controller/reconciler.rs [hunk 4]
```diff
@@ -248,21 +252,52 @@ impl ResourceContext {
...
+        for pod in pods.list(&lp).await? {
+            match pod.status {
+                Some(status) => {
+                    match status.conditions {
+                        Some(conditions) => {
+                            for condition in conditions {
+                                if condition.type_.eq("Ready") && condition.status.eq("True") {
+                                    continue
+                                } else {
+                                    warn!("Couldn't verify the ready condition of io-engine Pod '{}' in namespace '{}' to be true", pod.name_any(), pod.namespace().unwrap());
```
[the remark this answers is not in the document]
Comment (niladrih) [reply operators/src/upgrade/controller/reconciler.rs]: Added a closure.

File: operators/src/upgrade/controller/reconciler.rs [hunk 5]
```diff
@@ -235,6 +235,10 @@ impl ResourceContext {
                 debug!(?status, "Deleted collection of pods: status");
             }
         }
+
+        // Give kubernetes reconciler enough time to delete the pods.
+        tokio::time::sleep(Duration::from_secs(5)).await;
```
[the remark this answers is not in the document]
Comment (niladrih) [reply operators/src/upgrade/controller/reconciler.rs]: Refactored this section, added a loop.

**Comment by niladrih** [comment]
bors try

**Comment by niladrih** [comment]
bors try

**Comment by niladrih** [comment]
bors try

**Comment by niladrih** [comment]
bors try

**Comment by niladrih** [comment]
bors try

**Comment by niladrih** [comment]
bors try

**Comment by niladrih** [comment]
bors try

File: operators/Cargo.toml [hunk 6]
```diff
@@ -21,7 +21,15 @@ schemars = "0.8.10"
...
 serde_json = "1.0.83"
 serde_yaml = "0.9.9"
-tokio = {version="1.20.1", features = ["macros", "rt-multi-thread", "time"]}
+
+# The tokio crate version should always match the one used in actix-web.
+# The upgrade controller is deployed as a tokio task, and the actix-web must be able to
+# use the same runtime for its listener and workers.
+# E.g., the tokio version used to build actix-web 4.2.1 is tokio 1.13.1.
+# Ref: https://github.com/actix/actix-web/blob/web-v4.2.1/actix-web/Cargo.toml#L121
+tokio = { version = "1.13.1", features = ["rt-multi-thread", "macros", "time"] }
```
Comment (tiagolobocastro) [inline operators/Cargo.toml:30]: hmm this should not be true, only needs to be semver compatible which 1.20 is. In the control-plane we use 1.20 with no issues. What issues did you encounter?

File: operators/src/upgrade/config.rs [hunk 7]
```diff
@@ -8,23 +8,23 @@ use url::Url;
...
 
     /// The namespace we are supposed to operate in.
     #[clap(short, long, default_value = "mayastor")]
     namespace: String,
 
     /// The chart_name we are supposed to operate in.
-    #[clap(short, long, default_value = "mayastor")]
-    chart_name: String,
+    #[clap(short = 'r', long = "helm-release", default_value = "mayastor")]
+    helm_chart_release_name: String,
```
Comment (tiagolobocastro) [inline operators/src/upgrade/config.rs:27]: nit: release_name ?

File: operators/src/upgrade/controller/reconciler.rs [hunk 8]
```diff
@@ -216,15 +221,32 @@ impl ResourceContext {
...
+        for pod in initial_io_engine_pod_list.iter() {
+            match pod.metadata.uid.as_ref() {
+                Some(uid) => {
+                    // Building a set of io-engine Pod uid values. These values may be used to check
+                    // for io-engine Pod restarts.
+                    initial_io_engine_pod_uids.insert(uid);
+                }
+                None => {
+                    // It's ok to panic here because all kubernetes objects should have uids.
+                    panic!("Could not list all io-engine Pods' metadata.uids before restart");
```
Comment (tiagolobocastro) [inline operators/src/upgrade/controller/reconciler.rs:242]: Nit: it's fine but since this functions has Result<_,Error> we could still return an Error though?

File: operators/src/upgrade/helm/client.rs [hunk 9]
```diff
@@ -292,4 +298,16 @@ impl HelmClient {
             String::from_utf8(output.stdout).map_err(|error| Error::Utf8 { source: error })?;
         Ok(output)
     }
+    /*
+       pub(crate) fn show_chart(&mut self) -> ChartData {
```
Comment (tiagolobocastro) [inline operators/src/upgrade/helm/client.rs:302]: can remove this ?

File: k8s/plugin/src/resources/objects.rs [hunk 10]
```diff
@@ -233,9 +237,12 @@ pub(crate) fn upgrade_operator_deployment(
     upgrade_image: String,
     release_name: String,
 ) -> Deployment {
-    let rest_endpoint_arg = format!("--rest-endpoint={}", RestClient::get_or_panic().uri());
-    let namespace_arg = format!("--namespace={}", CliArgs::args().namespace);
-    let chart_name_arg = format!("--chart-name={}", &release_name);
+    let rest_endpoint_clusterip_url = format!("http://{}-api-rest:8081", &release_name);
```
[the remark this answers is not in the document]
Comment (niladrih) [reply k8s/plugin/src/resources/objects.rs:240]: No, it's not. The uri is being generated in the [build()](https://github.com/openebs/mayastor-extensions/blob/52a6888dd3bf784d1091deaa3dd12760018018e0/k8s/plugin/src/main.rs#L202) method, and only being used for forwarding.

File: operators/src/upgrade/config.rs (hunk 7 above)
Comment (niladrih) [reply operators/src/upgrade/config.rs:27]: Ok!

File: operators/src/upgrade/controller/reconciler.rs (hunk 8 above)
Comment (niladrih) [reply operators/src/upgrade/controller/reconciler.rs:242]: Yes, okay, done!

File: operators/src/upgrade/helm/client.rs (hunk 9 above)
Comment (niladrih) [reply operators/src/upgrade/helm/client.rs:302]: Yes.. ✔️

File: operators/Cargo.toml (hunk 6 above)
Comment (niladrih) [reply operators/Cargo.toml:30]: That's right. I misread this SO answer and wrote that up --> https://stackoverflow.com/questions/66416174/using-actix-from-a-tokio-app-mixing-actix-webmain-and-tokiomain. Thanks!

Comment (niladrih) [reply operators/Cargo.toml]: Tested and verified that this works with 1.20.

**Comment by niladrih** [comment]
bors try

File: operators/Cargo.toml [hunk 11]
```diff
@@ -21,7 +21,14 @@ schemars = "0.8.10"
 serde = "1.0.143"
 serde_json = "1.0.83"
 serde_yaml = "0.9.9"
-tokio = {version="1.20.1", features = ["macros", "rt-multi-thread", "time"]}
+
+# The tokio crate version should be compatible with the one used in actix-web.
+# The upgrade controller is deployed as a tokio task, and the actix-web must be able to
+# use the same runtime for its listener and workers.
+# E.g., the tokio version used to build actix-web 4.2.1 is tokio 1.13.1.
```
Comment (tiagolobocastro) [inline operators/Cargo.toml:28]: we can remove this comment as this is generally true for all crates with shared deps.
Also the way we set dependencies with the default resolver you're not guaranteed to get this exact version btw, it will be the "best for all"

File: operators/src/upgrade/controller/reconciler.rs [hunk 12]
```diff
@@ -216,15 +221,35 @@ impl ResourceContext {
...
+            match pod.metadata.uid.as_ref() {
+                Some(uid) => {
+                    // Building a set of io-engine Pod uid values. These values may be used to check
+                    // for io-engine Pod restarts.
+                    initial_io_engine_pod_uids.insert(uid);
+                }
+                None => {
+                    // All kubernetes objects should have uids.
+                    error!("Could not list all io-engine Pods' metadata.uids before restart");
+                    return Err(Error::ReconcileError {
```
Comment (tiagolobocastro) [inline operators/src/upgrade/controller/reconciler.rs:243]: Ah what I meant was add a more appropriate error for this situation so that you could match on it and decide to either retry or fail at that top level.

File: operators/src/upgrade/controller/reconciler.rs [hunk 13]
```diff
@@ -235,7 +260,28 @@ impl ResourceContext {
                 debug!(?status, "Deleted collection of pods: status");
             }
         }
-        Ok(Action::await_change())
+        debug!("Deleted io-engine Pods");
+
+        debug!("Waiting for new io-engine Pods to start...");
```
Comment (tiagolobocastro) [inline operators/src/upgrade/controller/reconciler.rs:265]: We're deleting all io-engine pods at once here, which means all volumes will be broken. Should we check that there are no published volumes?

File: operators/src/upgrade/helm/client.rs [hunk 14]
```diff
@@ -6,6 +6,7 @@ use std::{
 };
 
 use crate::upgrade::common::{constants::chart_dir_path, error::Error};
+//use crate::upgrade::helm::chart_data::ChartData;
```
Comment (tiagolobocastro) [inline operators/src/upgrade/helm/client.rs:9]: remove this line

File: operators/src/upgrade/k8s/crd/v0.rs [hunk 15]
```diff
@@ -80,15 +80,15 @@ impl UpgradeActionSpec {
 /// 'UpgradePhase' defines the status of each components.
 #[derive(Deserialize, Serialize, Clone, Debug, Eq, PartialEq, JsonSchema)]
 pub enum UpgradePhase {
-    /// Components in Waiting phase.
+    // Components in Waiting phase.
```
Comment (tiagolobocastro) [inline operators/src/upgrade/k8s/crd/v0.rs:83]: Why did you remove the doc comment?

**Comment by niladrih** [comment]
> These are the arguments for the operator itself right?
I wonder if we can define the operator via helm (and then the api-rest and what not is filled in via the template) and then the plugin would simply "apply" that.
Maybe this would solve some "issues", example: we're hardcoding RUST_LOG to info and image_pull_policy to Always, might be nice if we could tweak those via helm the same way we do for the other components.

One alternative could be to do something like... the user adds `--dry-run -o yaml` to the install command to get the yaml manifests. Then he can edit and apply it with configuration that he has validated.

**Comment by tiagolobocastro** [comment]
> > These are the arguments for the operator itself right?
> > I wonder if we can define the operator via helm (and then the api-rest and what not is filled in via the template) and then the plugin would simply "apply" that.
> > Maybe this would solve some "issues", example: we're hardcoding RUST_LOG to info and image_pull_policy to Always, might be nice if we could tweak those via helm the same way we do for the other components.
> 
> One alternative could be to do something like... the user adds `--dry-run -o yaml` to the install command to get the yaml manifests. Then he can edit and apply it with configuration that he has validated.

We could generate the yaml at build time, and then the operator simply has to apply it.
Maybe we an discuss this as a future enhancement (not for this PR)

File: operators/src/upgrade/k8s/crd/v0.rs (hunk 15 above)
Comment (niladrih) [reply operators/src/upgrade/k8s/crd/v0.rs:83]: Because, the kube-rs library uses the doc comment as input for the 'description' field in the CRD openapi spec json/yaml. That description is not openapi3 compliant, and leads to an invalid openapi3 spec.

**Comment by niladrih** [comment]
bors try

File: operators/src/upgrade/k8s/crd/v0.rs (hunk 15 above)
Comment (tiagolobocastro) [reply operators/src/upgrade/k8s/crd/v0.rs:83]: oh I see.. what would it take to make it compliant, have description would actually be good to have?

File: k8s/plugin/src/constant.rs [hunk 16]
```diff
@@ -40,7 +40,7 @@ pub(crate) const UPGRADE_OPERATOR_SERVICE_PORT: i32 = 8080;
 /// Service internal port constant for upgrade operator.
 pub(crate) const UPGRADE_OPERATOR_INTERNAL_PORT: i32 = 8080;
 /// Upgrade image tag
-pub(crate) const UPGRADE_IMAGE: &str = "openebs/mayastor-operator-upgrade:develop";
+pub(crate) const UPGRADE_IMAGE: &str = "niladrih/mayastor-operator-upgrade:final3";
```
Comment (tiagolobocastro) [inline k8s/plugin/src/constant.rs:43]: please revert this: probably better to add a cli option or something like that to avoid merging the wrong thing

File: operators/src/upgrade/controller/utils.rs [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            .get_volumes(max_entries, None, starting_token)
+            .await
+        {
+            Ok(vols) => {
+                let v = vols.into_body();
+                volumes.extend(v.entries);
+                starting_token = v.next_token;
+            }
+            Err(error) => {
+                panic!("Failed to list volumes. error: {}", error);
```
Comment (tiagolobocastro) [inline operators/src/upgrade/controller/utils.rs:82]: why is this a panic, this should be handled as errors via rest may occur.
This fn should return Result.

File: k8s/plugin/src/resources/objects.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (niladrih) [reply k8s/plugin/src/resources/objects.rs:240]: ImagePullPolicy, image registry, repository and tag has to be customizable. I'll take up this change in a separate PR.

File: k8s/plugin/src/constant.rs (hunk 16 above)
Comment (niladrih) [reply k8s/plugin/src/constant.rs:43]: Oops

Comment (niladrih) [reply k8s/plugin/src/constant.rs]: Fixed it, I'm not adding a cli flag for this at the moment.

File: operators/src/upgrade/controller/utils.rs (hunk 17 above)
Comment (niladrih) [reply operators/src/upgrade/controller/utils.rs]: Added an error return for this.

**Comment by niladrih** [comment]
bors try

**Comment by niladrih** [comment]
bors try

**Review by tiagolobocastro — APPROVED** [review]

**Comment by niladrih** [comment]
bors merge

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by niladrih** [comment]
bors merge
